akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ProducerController.scala

```diff
@@ -0,0 +1,257 @@
+ * used with a [[DurableProducerQueue]]. Then the unconfirmed messages are stored in a durable way so
+ * that they can be redelivered when the producer is started again. An implementation of the
+ * `DurableProducerQueue` is provided by `EventSourcedProducerQueue` in `akka-persistence-typed`.
+ *
+ * Instead of using `tell` with the `sendNextTo` in the `RequestNext` the producer can use `context.ask`
+ * with the `askNextTo` in the `RequestNext`. The difference is that a reply is sent back when the
+ * message has been handled. If a `DurableProducerQueue` is used then the reply is sent when the message
+ * has been stored successfully, but it might not have been processed by the consumer yet. Otherwise the
+ * reply is sent after the consumer has processed and confirmed the message.
+ *
+ * If the consumer crashes a new `ConsumerController` can be connected to the original `ProducerConsumer`
+ * without restarting it. The `ProducerConsumer` will then redeliver all unconfirmed messages.
+ *
+ * It's also possible to use the `ProducerController` and `ConsumerController` without resending
+ * lost messages, but the flow control is still used. This can for example be useful when both consumer and
+ * producer are know to be located in the same local `ActorSystem`. This can be more efficient since messages
+ * don't have to be kept in memory in the `ProducerController` until they have been
+ * confirmed, but the drawback is that lost messages will not be delivered. See configuration
+ * `only-flow-control` of the `ConsumerController`.
+ */
+@ApiMayChange // TODO when removing ApiMayChange consider removing `case class` for some of the messages
+object ProducerController {
+  import ProducerControllerImpl.UnsealedInternalCommand
+
+  type SeqNr = Long
+
+  sealed trait Command[A] extends UnsealedInternalCommand
+
+  /**
+   * Initial message from the producer actor. The `producer` is typically constructed
+   * as a message adapter to map the [[RequestNext]] to the protocol of the producer actor.
+   *
+   * If the producer is restarted it should send a new `Start` message to the
+   * `ProducerController`.
+   */
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  /**
+   * The `ProducerController` sends `RequestNext` to the producer when it is allowed to send one
+   * message via the `sendNextTo` or `askNextTo`. Note that only one message is allowed, and then
+   * it must wait for next `RequestNext` before sending one more message.
+   */
+  final case class RequestNext[A](
+      producerId: String,
+      currentSeqNr: SeqNr,
+      confirmedSeqNr: SeqNr,
+      sendNextTo: ActorRef[A],
+      askNextTo: ActorRef[MessageWithConfirmation[A]])
+
+  /**
+   * For sending confirmation message back to the producer when the message has been confirmed.
+   * Typically used with `context.ask` from the producer.
+   *
+   * If `DurableProducerQueue` is used the confirmation reply is sent when the message has been
```

Comment:
That's an interesting consideration. My thinking is that you would use a durable queue for safe (quick) handoff. Once it's stored it will eventually be handled. That ack is more interesting to use as signal of when to return OK to a HTTP request. If the full processing was quick the durable queue wouldn't be needed, since it could then be a responsibility of the original client to resend.

For the ProducerController it's possible to see the processing confirmations in the `RequestNext` message by correlating sequence numbers, but that is hidden for Sharding and WorkPulling.